File: SmogWatch WatchKit Extension/ComplicationController.m
//
//  ComplicationController.m
//  SmogWatch WatchKit Extension
//
//  Created by Kuba Suder on 27.12.2018.
//  Copyright © 2018 Kuba Suder. Licensed under WTFPL license.
//

#import "ComplicationController.h"
#import "DataStore.h"

static NSTimeInterval MeasurementValidityTime = 3600 * 3;
static NSString *PMTitleText = @"PM10";
static NSString *PMShortTitleText = @"PM";
static NSString *PlaceholderText = @"–";
static NSString *SampleValueText = @"50";

@interface ComplicationController () {
    DataStore *dataStore;
}

@end

@implementation ComplicationController

- (instancetype)init {
    self = [super init];
    if (self) {
        dataStore = [[DataStore alloc] init];
    }
    return self;
}


#pragma mark - Timeline Configuration

- (void)getSupportedTimeTravelDirectionsForComplication:(CLKComplication *)complication
                                            withHandler:(void(^)(CLKComplicationTimeTravelDirections directions))handler {
    handler(0);
}

- (void)getTimelineEndDateForComplication:(CLKComplication *)complication
                              withHandler:(void(^)(NSDate * __nullable date))handler {
    handler([[dataStore lastMeasurementDate] dateByAddingTimeInterval:MeasurementValidityTime]);
}

#pragma mark - Timeline Population

- (void)getCurrentTimelineEntryForComplication:(CLKComplication *)complication
                                   withHandler:(void(^)(CLKComplicationTimelineEntry * __nullable))handler {
    CLKComplicationTimelineEntry *entry;

    if (dataStore.lastMeasurementDate && dataStore.currentLevel) {
        NSString *valueText = [NSString stringWithFormat:@"%d", (NSInteger) round(dataStore.currentLevel)];

        CLKComplicationTemplateModularSmallStackText *template = [[CLKComplicationTemplateModularSmallStackText alloc] init];
        template.line1TextProvider = [CLKSimpleTextProvider textProviderWithText:PMTitleText shortText:PMShortTitleText];
        template.line2TextProvider = [CLKSimpleTextProvider textProviderWithText:valueText];

        entry = [CLKComplicationTimelineEntry entryWithDate:dataStore.lastMeasurementDate complicationTemplate:template];
    } else {
        CLKComplicationTemplateModularSmallStackText *template = [[CLKComplicationTemplateModularSmallStackText alloc] init];
        template.line1TextProvider = [CLKSimpleTextProvider textProviderWithText:PMTitleText shortText:PMShortTitleText];
        template.line2TextProvider = [CLKSimpleTextProvider textProviderWithText:PlaceholderText];

        entry = [CLKComplicationTimelineEntry entryWithDate:[NSDate date] complicationTemplate:template];
    }

    handler(entry);
}


#pragma mark - Placeholder Templates

- (void)getLocalizableSampleTemplateForComplication:(CLKComplication *)complication
                                        withHandler:(void(^)(CLKComplicationTemplate * __nullable complicationTemplate))handler {
    // This method will be called once per supported complication, and the results will be cached
    CLKComplicationTemplateModularSmallStackText *template = [[CLKComplicationTemplateModularSmallStackText alloc] init];
    template.line1TextProvider = [CLKSimpleTextProvider textProviderWithText:PMTitleText shortText:PMShortTitleText];
    template.line2TextProvider = [CLKSimpleTextProvider textProviderWithText:SampleValueText];

    handler(template);
}

@end

File: SmogWatch WatchKit Extension/ComplicationController.swift
//
//  ComplicationController.swift
//  SmogWatch WatchKit Extension
//
//  Created by Kuba Suder on 15.12.2018.
//  Copyright © 2018 Kuba Suder. Licensed under WTFPL license.
//

import ClockKit
import os.log

private let MeasurementValidityTime = 3600 * 3
private let log = OSLog(subsystem: OSLog.subsystem, category: "Complication Controller")

class ComplicationController: NSObject, CLKComplicationDataSource {

    let dataStore = DataStore()


    // MARK: - Timeline Configuration

    func getSupportedTimeTravelDirections(
        for complication: CLKComplication,
        withHandler handler: @escaping (CLKComplicationTimeTravelDirections) -> Void)
    {
        os_log("getSupportedTimeTravelDirections() for complication %{public}@", log: log,
               complication.family.description);
        handler([])
    }


    // MARK: - Timeline Population

    func getCurrentTimelineEntry(
        for complication: CLKComplication,
        withHandler callback: @escaping (CLKComplicationTimelineEntry?) -> Void)
    {
        os_log("getCurrentTimelineEntry() for complication %{public}@", log: log, complication.family.description)

        let entry: CLKComplicationTimelineEntry
        let handler = ComplicationHandlers.handler(for: complication)

        if let date = dataStore.lastMeasurementDate, let level = dataStore.currentLevel {
            let template = handler.template(for: level)
            entry = CLKComplicationTimelineEntry(date: date, complicationTemplate: template)
            os_log("getCurrentTimelineEntry() -> %@ %.0f", log: log, date as NSDate, level)
        } else {
            let template = handler.templateForNoValue()
            entry = CLKComplicationTimelineEntry(date: Date(), complicationTemplate: template)
            os_log("getCurrentTimelineEntry() -> %@ n/a", log: log, NSDate())
        }

        callback(entry)
    }


    // MARK: - Placeholder Templates

    func getLocalizableSampleTemplate(
        for complication: CLKComplication,
        withHandler callback: @escaping (CLKComplicationTemplate?) -> Void)
    {
        os_log("getLocalizableSampleTemplate() for complication %{public}@", log: log, complication.family.description)

        // This method will be called once per supported complication, and the results will be cached

        let handler = ComplicationHandlers.handler(for: complication)
        let template = handler.templateForSampleValue()

        callback(template)
    }
}
